src/contrasts.jl

```diff
@@ -368,25 +368,144 @@ function contrasts_matrix(C::HelmertCoding, baseind, n)
+between levels ``n`` and ``n+1`` is zero.
+
+# Examples
+
+```jldoctest seqdiff
+julia> seqdiff = StatsModels.ContrastsMatrix(SeqDiffCoding(), ["a", "b", "c", "d"]).matrix
+4×3 Array{Float64,2}:
+ -0.75  -0.5  -0.25
+  0.25  -0.5  -0.25
+  0.25   0.5  -0.25
+  0.25   0.5   0.75
+```
+
+The interpretation of sequential difference coding may be hard to see from the
+contrasts matrix itself.  The corresponding hypothesis matrix shows a clearer
+picture.  From the rows of the hypothesis matrix, we can see that these
+contrasts test the difference between the first and second levels, the second
+and third, and the third and fourth, respectively:
+
+```jldoctest seqdiff
+julia> round.(pinv(seqdiff), digits=2)
+3×4 Array{Float64,2}:
+ -1.0   1.0  -0.0   0.0
+ -0.0  -1.0   1.0  -0.0
+  0.0  -0.0  -1.0   1.0
+```
+
+"""
+SeqDiffCoding
+
+function contrasts_matrix(C::SeqDiffCoding, baseind, n)
+    mat = zeros(n, n-1)
+    for col in 1:n-1
+        mat[1:col, col] .= col-n
+        mat[col+1:end, col] .= col
+    end
+    return mat ./ n
+end
+
+
+"""
+    HypothesisCoding(hypotheses::Matrix[, levels]])
+
+Specify how to code a categorical variable in terms of a *hypothesis matrix*.
+For a variable with ``k`` levels, this should be a ``k-1 \times k`` matrix.
+Each row of the hypothesis corresponds to the hypothesis tested by the
+corresponding predictor, given as the weights given to the "cell mean" of each
+of the ``k`` values of the predictor.
+
+For instance, if we have a variable which has four levels A, B, C, and D, and we
+want to test the hypothesis that the difference between the average outcome for
+levels A and B is different from zero, the corresponding row of the hypothesis
+matrix would be `[-1, 1, 0, 0]`.  Likewise, to test whether the difference
+between B and C is different from zero, the hypothesis vector would be `[0, -1,
+1, 0]`.  To test each "successive difference" hypothesis, the full hypothesis
+matrix would be
+
+```jldoctest hyp
+julia> sdiff_hypothesis = [-1  1  0  0
+                            0 -1  1  0
+                            0  0 -1  1];
+```
+
+Contrasts are derived the hypothesis matrix by taking the pseudoinverse:
+
+```jldoctest hyp
+julia> sdiff_contrasts = pinv(sdiff_hypothesis)
+4×3 Array{Float64,2}:
+ -0.75  -0.5  -0.25
+  0.25  -0.5  -0.25
+  0.25   0.5  -0.25
+  0.25   0.5   0.75
+```
+
+Which is what is produced by constructing a [`ContrastsMatrix`](@ref) from a
+`HypothesisCoding` instance:
+
+```jldoctest hyp
+julia> StatsModels.ContrastsMatrix(HypothesisCoding(sdiff_hypothesis), ["a", "b", "c", "d"]).matrix
+4×3 Array{Float64,2}:
+ -0.75  -0.5  -0.25
+  0.25  -0.5  -0.25
+  0.25   0.5  -0.25
+  0.25   0.5   0.75
+```
+
+
+"""
+mutable struct HypothesisCoding <: AbstractContrasts
+    hypotheses::Matrix
+    contrasts::Matrix
+    base::Nothing
+    levels::Union{Vector,Nothing}
+
+    function HypothesisCoding(hypotheses, base, levels)
```

Comment:
Why specify `base` if it's always `nothing`?

Reply:
I think there's code that tries to access the field.  I could overload getproperty for this type instead though, or refactor that code...

Reply:
Since `HypothesisCoding` is new, how could existing code try to access that field? :-)

 Anyway, in general I think fields are considered private unless specified otherwise. And unexporting `ContrastsCoding` is much more breaking. BTW, maybe we should keep exporting it for now until we want to make another breaking release? It's simpler to break many things at the same time than force people to update small things continuously.

Reply:
Sounds good.  Let's keep it exported but maybe add a warning to teh constructor or something?

Reply:
I added a depwarn because the deprecations weren't quite doing it correctly (giving a spurious method overwritten warning)

Reply:
The base field is used in the constructor of `ContrastsMatrix`.  Could change that to use an accessor method but that seems like overkill at this point.